File: src/main/java/com/hileco/cortex/analysis/GraphBuilder.java
package com.hileco.cortex.analysis;

import com.hileco.cortex.analysis.processors.Processor;
import com.hileco.cortex.instructions.Instruction;
import lombok.Value;

import java.util.List;

@Value
public class GraphBuilder {
    private List<Processor> processors;

    public Graph build(List<Instruction> instructions) {
        var graph = new Graph(instructions);
        this.processors.forEach(processor -> processor.process(graph));
        return graph;
    }
}

File: src/main/java/com/hileco/cortex/analysis/GraphBuilder.kt
package com.hileco.cortex.analysis

import com.hileco.cortex.analysis.processors.*
import com.hileco.cortex.instructions.Instruction

class GraphBuilder(val processors: List<Processor>) {
    fun build(instructions: List<Instruction>): Graph {
        val graph = Graph(instructions)
        processors.forEach { processor -> processor.process(graph) }
        return graph
    }

    companion object {
        val BASIC_GRAPH_BUILDER = GraphBuilder(listOf(
                ParameterProcessor(),
                FlowProcessor()
        ))

        val OPTIMIZED_GRAPH_BUILDER = GraphBuilder(listOf(
                ParameterProcessor(),
                FlowProcessor(),

                KnownProcessor(),
                KnownJumpIfProcessor(),
                KnownLoadProcessor(mapOf()),

                DeadStartProcessor(),
                DeadEndProcessor(),
                DeadLoadProcessor(),
                DeadSaveProcessor(),
                DeadSwapProcessor(),

                JumpIllegalProcessor(),
                JumpUnreachableProcessor(),
                JumpThreadingProcessor(),
                InstructionHoistProcessor(),
                InliningProcessor(),

                ParameterProcessor(),
                FlowProcessor(),

                CompositeConstraintProcessor(),
                DeadPathConstraintProcessor()
        ))
    }
}
